File: gateway/src/services/ai/tools/codeDiff.ts
import { diffLines, diffChars } from 'diff';
import { CodeDiffParams, CodeDiffResult } from '@/types/tools.js';

export const generateCodeDiff = async (params: CodeDiffParams): Promise<CodeDiffResult> => {
  const { oldCode, newCode, language } = params;
  
  // Use line-based diff for better readability
  const lineDiffs = diffLines(oldCode, newCode);
  
  const diffs: CodeDiffResult['diffs'] = [];
  let currentLineNumber = 1;
  let linesAdded = 0;
  let linesRemoved = 0;
  let linesChanged = 0;

  for (const part of lineDiffs) {
    const lines = part.value.split('\n').filter(line => line !== '');
    
    if (part.added) {
      linesAdded += lines.length;
      lines.forEach(line => {
        diffs.push({
          type: 'added',
          content: line,
          lineNumber: currentLineNumber++
        });
      });
    } else if (part.removed) {
      linesRemoved += lines.length;
      lines.forEach(line => {
        diffs.push({
          type: 'removed',
          content: line,
          lineNumber: currentLineNumber
        });
      });
      // Don't increment line number for removed lines
    } else {
      // Unchanged lines
      lines.forEach(line => {
        diffs.push({
          type: 'unchanged',
          content: line,
          lineNumber: currentLineNumber++
        });
      });
    }
  }

  // Calculate changed lines (lines that were both removed and added)
  linesChanged = Math.min(linesAdded, linesRemoved);

  return {
    diffs,
    summary: {
      linesAdded,
      linesRemoved,
      linesChanged
    }
  };
};

File: ai-services/src/services/tools/code_diff.py
"""Code diff generation service."""
import difflib
from typing import List

from src.models.tools import CodeDiffParams, CodeDiffResult, DiffLine, DiffSummary


class CodeDiffService:
    """Service for generating code diffs."""

    async def generate_diff(self, params: CodeDiffParams) -> CodeDiffResult:
        """Generate code diff between old and new code."""
        old_lines = params.old_code.splitlines(keepends=True)
        new_lines = params.new_code.splitlines(keepends=True)
        
        # Generate unified diff
        diff = difflib.unified_diff(
            old_lines,
            new_lines,
            fromfile="old_code",
            tofile="new_code",
            lineterm=""
        )
        
        # Parse diff and create structured result
        diffs: List[DiffLine] = []
        lines_added = 0
        lines_removed = 0
        current_line_number = 1
        
        # Use ndiff for better line-by-line comparison
        ndiff = difflib.ndiff(old_lines, new_lines)
        
        for line in ndiff:
            if line.startswith("  "):  # Unchanged line
                diffs.append(DiffLine(
                    type="unchanged",
                    content=line[2:].rstrip("\n"),
                    line_number=current_line_number
                ))
                current_line_number += 1
            elif line.startswith("- "):  # Removed line
                diffs.append(DiffLine(
                    type="removed",
                    content=line[2:].rstrip("\n"),
                    line_number=current_line_number
                ))
                lines_removed += 1
            elif line.startswith("+ "):  # Added line
                diffs.append(DiffLine(
                    type="added",
                    content=line[2:].rstrip("\n"),
                    line_number=current_line_number
                ))
                lines_added += 1
                current_line_number += 1
            elif line.startswith("? "):  # Diff hint - skip
                continue
        
        # Calculate changed lines (minimum of added and removed)
        lines_changed = min(lines_added, lines_removed)
        
        summary = DiffSummary(
            lines_added=lines_added,
            lines_removed=lines_removed,
            lines_changed=lines_changed
        )
        
        return CodeDiffResult(diffs=diffs, summary=summary)


# Global service instance
code_diff_service = CodeDiffService()